### PR TITLE
Add the documentation for the Asynchronous API usages

### PR DESCRIPTION
**Pull request by ihsandemir**
Added the documentation for the asynchronous API usage.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1329,6 +1332,68 @@ A Map usage example is shown below.
     map.replace("key", "value", "newvalue");
 ```
 
+#### 7.4.1.1. Using IMap Non-Blocking Async Methods
+Hazelcast IMap provides asynchronous methods for more powerful operations like batched writing or batched reading. Asynchronous methods do not block but return immediately with an `ICompletableFuture` interface. You can later query the result of the operation using this interface or you can even provide a callback method to be executed on the user executor threads when the response to the call is received.
```
Comment (Serdaro) [inline README.md]: ... batched reading. The asynchronous methods ...

File: README.md [hunk 2]
```diff
@@ -1329,6 +1332,68 @@ A Map usage example is shown below.
...
+    
+    // later on get the result of the put operation
+    boost::shared_ptr<std::string> result = future->get();
+    if (result.get()) {
+        std::cout << "There was a previous value for key. The value was:" << *result << std::endl;
+    } else {
+        std::cout << "There was no previous value for key." << std::endl;
+    }
+```
+The asynchronous methods always return an `ICompletableFuture` pointer. The method does not block for IO. You can get the result later with the `ICompletableFuture::get()` method. `ICompletableFuture::get()` blocks until the response is received or an exception occurs.
```
Comment (Serdaro) [inline README.md]: ... pointer. The method does not block IO communication. You can ...

File: README.md [hunk 3]
```diff
@@ -1329,6 +1332,68 @@ A Map usage example is shown below.
...
+The asynchronous methods always return an `ICompletableFuture` pointer. The method does not block for IO. You can get the result later with the `ICompletableFuture::get()` method. `ICompletableFuture::get()` blocks until the response is received or an exception occurs.  
+
+Other asynchronous `IMap` methods are:
+ 
+- `IMap::getAsync`
+- `IMap::removeAsync`
+- `IMap::setAsync`
+- `IMap::getAsync`
+
+You can also provide a callback instance to the future using the `andThen` method. The callback is executed upon reception of the call response or exception and you can put your code into the callback methods. The callback method will be executed in the user executor thread which is different from the user's main program thread and hence it will not block the user. An example callback usage is shown below:
```
Comment (Serdaro) [inline README.md]: ... executed upon the reception of the call response or exception, and you can ...

File: README.md [hunk 4]
```diff
@@ -1558,6 +1665,61 @@ An Atomic Long usage example is shown below.
     std::cout << "Count is:" << counter.get() << std::endl; // Counter is 1000000
 ```
 
+#### 7.4.10.1. Using Atomic Long Non-Blocking Async Methods
+IAtomicLong provides asynchronous methods where you can execute a bunch of non-blocking methods and get the response later in your code or just let the atomic long operations complete at the background. You can also use callbacks to be executed upon the completion of the atomic long operations.
```
Comment (Serdaro) [inline README.md]: Hazelcast IAtomicLong provides ... _(for the sake of consistency with the other newly added sections' intros)._

File: README.md [hunk 5]
```diff
@@ -1558,6 +1665,61 @@ An Atomic Long usage example is shown below.
...
+    // Do some other work
+
+    // Get the result of the incrementAndGetAsync api using the future
+    boost::shared_ptr<int64_t> result = future->get();
+
+    // It will print the value as 1
+    std::cout << "The counter value is " << *result << std::endl;
+```
+
+Other asynchronous methods provided by IAtomicLong structure are:
```
Comment (Serdaro) [inline README.md]: ... by the IAtomicLong structure ...

File: README.md (hunk 2 above)
Comment (ihsandemir) [reply README.md]: changed to `does not block on IO thread`

**Pull request merged**